### PR TITLE
refactor: separate precompile verification from proving

### LINKED ISSUE #2282: precompiles: consider a separate precompile crate
# Summary

Create a new top-level workspace crate:

'precompiles/'

with package name 'miden-precompiles'.

This crate should become the first-class home for Precompile VM (PVM) work inside 'miden-vm'.

The immediate goal is to create a clean landing zone for existing private/research PVM work so it can be moved into this repo incrementally, reviewed in smaller pieces, and developed on a focused tracking branch before merging into 'next'.

# Why

PVM work is not just "add another chip". It cuts across several parts of the system:

- deferred operation semantics;
- PVM AIR constraints;
- PVM trace layout and witness generation;
- core PVM chips;
- concrete precompile implementations;
- Rust-side handlers and registries;
- MASM wrappers and library exposure;
- tests that check the whole path end-to-end.

Without a dedicated crate, this work tends to spread across 'core', 'processor', 'air', 'prover', 'verifier', and 'miden-core-lib'. That makes it harder to develop quickly and harder to review.

A dedicated 'miden-precompiles' crate gives us one place to develop the PVM subsystem while keeping the boundary with the base VM clear. It also makes it easier to maintain a PVM tracking branch that should not constantly conflict with unrelated VM/API refactors.

# Proposed Direction

Start with one crate: 'miden-precompiles'.

Initially, this crate should contain the PVM-specific implementation work:

- PVM AIR constraints;
- PVM trace layout;
- PVM witness generation;
- the eval chip;
- the chunks chip;
- the permutation chip;
- the permutation-controller chip;
- concrete precompile implementations as they become ready;
- a built-in registry of supported precompiles;
- end-to-end PVM tests.

We should not start by splitting this into many crates. A later split into something like 'miden-precompile-vm' and 'miden-precompile-lib' may make sense, but only after the first end-to-end path is working and we understand the dependency graph better.

# Boundary With The Base VM

For now, assume the generic deferred-state framework stays in 'miden-core'.

That includes shared VM ABI pieces such as:

- deferred tags and nodes;
- transcript/root shape;
- wire format;
- generic deferred-state data structures;
- generic registry interfaces.

The 'processor' should remain generic. It may carry a registry and update deferred state through system events, but it should not know which concrete precompiles exist.

The new 'miden-precompiles' crate should own the concrete PVM/precompile logic: the chips, constraints, trace/witness generation, and concrete precompile implementations.

# MASM Wrappers

MASM wrapper packaging does not need to be finalized in this issue.

For now, wrappers can continue to live in or be exposed through 'miden-core-lib' where that is practical. The new crate gives us a place to add first-class wrappers as concrete precompiles become ready, with Keccak likely being the first useful vertical slice.

Longer term, we should decide whether precompile wrappers belong in a separate 'miden::precompiles' library, are re-exported by 'miden-core-lib', or both.

# Relation To Current Work

Related work:

- #3170 adds the content-addressed deferred-DAG framework.
- #3172 cuts production precompiles over to the deferred-DAG proof model.
- #3176 adds native secp256k1 ECDSA support and useful field/scalar/group building blocks.

The PVM crate builds on this direction. It gives us a place to move the lower-level PVM infrastructure into the main repo first, then add concrete vertical slices such as Keccak and ECDSA-related field/group operations.

This issue should not require reorganizing #3176 before it lands. That PR can land independently and later serve as context for deciding which expensive native operations should become PVM-backed.

# First Milestones

1. Add top-level 'precompiles/' crate as 'miden-precompiles'.
2. Keep 'miden-core::deferred' as the shared generic framework for now.
3. Move the lower-level private PVM infrastructure into the new crate in reviewable chunks.
4. Add the base PVM AIR constraints, trace layout, and witness-generation machinery.
5. Add the core chips: eval, chunks, permutation, and permutation controller.
6. Add one small end-to-end test proving the crate boundary works.
7. Follow up with concrete precompile verticals, likely starting with Keccak and then ECDSA-related field/group operations.

# Open Questions

- Does the single-crate approach create dependency issues once PVM prover/verifier code is added?
- Which deferred-framework pieces, if any, should eventually move out of 'miden-core'?
- Should the built-in registry initially contain all supported precompiles?
- What should the long-term MASM library shape be?
- What is the first concrete vertical slice after the base PVM infrastructure lands?
- When should we split 'miden-precompiles' into smaller crates, if ever?

# Follow-Up Issues

- Eval-chip design.
- PVM trace/witness generation design.
- Keccak PVM vertical slice.
- ECDSA PVM milestone.
- Generic field-chip and group-chip design.
- MASM precompile library / namespace strategy.
- Possible future crate split.

<details>
I propose a new workspace crate (e.g., crates/precompiles) that centralizes all precompile behavior and types. It would define an explicit set of supported precompiles via an enum (each precompile gets a variant), and a single trait that encompasses both execution-time handling and verification (either by having the verifier inherit EventHandler, or by providing a unified trait for which we implement the event handler). The trait would expose the Event ID explicitly so the ID is shared and discoverable; we can also introduce a typed PrecompileId and a simple tag/version policy to reduce misuse and enable upgrades. 

Each precompile would live in one place and be required to provide
- a handler
- a verifier
- a MASM wrapper which safely emits the event calling the handler
- AirScript constraints and witness generation for the Precompile VM.
By default, the prover and verifier only support the enum-defined set in this crate

It remains technically possible, as an advanced feature, to add custom precompiles, but users must also provide a separate STARK verifier for them to be acceptable in production. We would remove the dynamic registry behavior and remove verify_with_precompiles, keeping a single verify wired to the built-in set and retaining the notion of a precompile ID (now explicitly surfaced by the trait). 

This crate is also the natural future home of the Precompile VM.

The main drawback is that we introduce another source/library of MASM code outside the stdlib. 
<details/>

### PR DESCRIPTION
**Pull request by huitseeker**
- `miden-verifier` now uses `miden-precompiles-verifier`. Shared AIR code is in `miden-precompiles-air`. 
- All proving work remains in `miden-precompiles-prover`: verifier users do not build prover code.

This completes the crate split discussed in [#2282](https://github.com/0xMiden/miden-vm/issues/2282). [#3222](https://github.com/0xMiden/miden-vm/pull/3222) created the runtime crate. [#3293](https://github.com/0xMiden/miden-vm/pull/3293) and [#3343](https://github.com/0xMiden/miden-vm/pull/3343) added the first PVM proof path. [#3467](https://github.com/0xMiden/miden-vm/pull/3467) added the MASM verifier input adapter. [#3672](https://github.com/0xMiden/miden-vm/pull/3672) settled the MASM package layout. The verifier split was [left until that work was complete](https://github.com/0xMiden/miden-vm/issues/2282#issuecomment-5079958774).

- Existing PVM proof bytes still verify when their registry root matches. 
- The relation digest, preprocessed commitment, ACE registry root, AIR order, and proof encoding are unchanged. 
- A pinned proof from before the split verifies with its recorded root and fails with a different root.

Closes #2282

### CODE REVIEW

**Review by Al-Kindi-0 — APPROVED** [review]
Looks good, modulo the change log entry

File: Cargo.toml [hunk 1]
```diff
@@ -36,6 +36,8 @@ members = [
     "crates/utils-sync",
     "miden-vm",
     "crates/precompiles",
+    "crates/precompiles-air",
+    "crates/precompiles-verifier",
     "crates/lib/core/codegen",
     "crates/precompiles-prover",
```
Comment (bobbinth) [inline Cargo.toml:42]: We'll need to set up trusted publishing for `precompiles-air` and `precompiles-verifier` crates.

Also question: is the reason `crates/lib/core/codegen` mixed in between the precompiles crates because it is related to precompiles?

Also also, after this PR, may be a good time to update the [Project structure](https://github.com/0xMiden/miden-vm#project-structure) section in the main README as it has become quite outdated.

Comment (huitseeker) [reply Cargo.toml:42]: 1. The current position of `crates/lib/core/codegen` is historical: it used to be `precompiles/codegen`, and I will move it next to `crates/lib/core`.
2. For miden-precompiles-air and miden-precompiles-verifier, recall you need to publish one version of each with an API token. Then you just need to add a trusted publisher for repository 0xMiden/miden-vm and workflow workspace-publish.yml, using the release environment.
